text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,626 @@
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. **Package Discovery:** The back-end for Construct Hub starts with a custom
+   built event source, which implements the logic necessary to detect new
+   packages that are relevant to the Construct Hub, and sends messages to an SQS
+   queue for further processing. Notifications have the following attributes:
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `assembly`  | The `.jsii` assembly included in the package
+   `time`      | The timestamp at which the version was created
+   `integrity` | An integirity check for the complete record
+
+   In cases where the `assembly` object is too large, it may be staged in S3
+   instead of being sent directly as part of the message payload. In such cases,
+   the `assembly` field contains an S3 URI instead of a JSON object.
+
+   Construct Hub comes with a built-in integration function that listents to
+   changes on the [npmjs.com registry](https://skimdb.npmjs.com/registry)
+   database. Operators of private instances are able to interface with other
+   package registries (including private ones), such as CodeArtifact
+   repositories, etc.
+
+1. **Ingestion:** A Lambda function then picks up messages from the SQS queue
+   and prepares the artifacts consumed by the front-end application, stored in a
+   dedicated S3 bucket using the following key format:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly.json`
+
+   This function validates the contents of the `assembly` to ensure the message
+   was well-formed. In case the message is found to be incoherent, it is sent to
+   a dead-letter queue and an alarm is triggered.
+
+   It also creates or updates the `assemblies/${assembly.name}/versions.json`
+   object, which contains a list of all versions ever indexed for the package,
+   matched to the versions' current status (`deprecated`, `latest`, ...):
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `v${major}` | All versions part of this major version stream, and their status
+
+   [SemVer]: https://semver.org/spec/v2.0.0.html
+
+   In order to ensure consistency of the contents of the `versions.json` object,
+   its content will always be generated from the live inventory of objects in S3
+   (and not incrementally using the update trigger event), and the last event
+   timestamp will be written in an object metadata attribute. The S3 Bucket will
+   have object versioning enabled, and once a new version of the `versions.json`
+   object has been created, its metadata will be checked against any previous
+   versions of the object, and the "current" version will be set to the one with
+   the latest event timestamp if needed.
```

Comment:
At some point I want to allow users to view any version of the package they want, and I'd rather avoid having to resort to an S3 List operation if possible. This file can also serve as a tool to diagnose system operations (what versions have been processed, etc....).

It is however not a priority for the MVP/DevPreview, so it could be considered a "nice-to-have". I'd kinda like to keep it in there however... Unless we all agree we don't want to design/scope this until we know we require it.